File: rust/src/solution1.rs
use crate::common::Solution;
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map = HashMap::new();
        let mut result = Vec::new();

        let length = nums.len();
        for i in 0..length {
            let num = nums[i];
            let lookup_num = target - num;
            match map.get(&lookup_num) {
                None => {
                    map.insert(num, i as i32);
                }
                Some(index) => {
                    result.push(*index);
                    result.push(i as i32);
                    return result;
                }
            }
        }
        return result;
    }
}

File: java/Solution1.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap();

        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            int lookup = target - num;
            Integer preLookupIndex = map.get(lookup);
            if (preLookupIndex != null) return new int[]{preLookupIndex, i};

            // trick -> this could easily handle duplicated case:
            //          [3, 2, 3]
            map.put(num, i);
        }
        return null; 
    }
}